File: DataStructures/UnionFind.js
"use strict";

class UnionFind1 {

    constructor(numNodes) {

        this.parents = [];
        this.ranks = [];
        this.groups = numNodes;

        for (let i = 0; i < numNodes; i++) {

            this.parents.push(i);
            this.ranks.push(0);

        }

    }

    find(node) {

        while (node != this.parents[node]) {

            node = this.parents[node];

        }

        return node;

    }

    union(a, b) {

        let aGroup = this.find(a),
            bGroup = this.find(b);

        if (aGroup != bGroup) {

            if (this.ranks[aGroup] > this.ranks[bGroup]) {

                this.parents[bGroup] = aGroup;

            } else if (this.ranks[bGroup] > this.ranks[aGroup]) {

                this.parents[aGroup] = bGroup;

            } else {

                this.parents[bGroup] = aGroup;

                this.ranks[aGroup]++;

            }

            this.groups--;

        }

    }

    isConnected(a, b) {

        return this.find(a) == this.find(b);

    }

}

for (let unionFindClass of [UnionFind1]) {

    let uf = new unionFindClass(10);

    console.log(uf.groups);

    uf.union(0, 1);
    uf.union(2, 3);
    uf.union(4, 5);

    console.log(uf.isConnected(0, 3));

    uf.union(1, 2);

    console.log(uf.isConnected(0, 3));

    console.log(uf.isConnected(0, 2));

    console.log(uf.isConnected(2, 5));

    console.log(uf.groups);

    console.log("------");

}

File: DataStructures/UnionFind.py

class UnionFind1:

    def __init__(self, num_nodes):

        self.parents = [ i for i in range(num_nodes) ]
        self.ranks = [ 0 ] * num_nodes
        self.groups = num_nodes

    def find(self, node):

        while node != self.parents[node]:

            node = self.parents[node]

        return node

    def union(self, a, b):

        a_group, b_group = self.find(a), self.find(b)

        if a_group != b_group:

            if self.ranks[a_group] > self.ranks[b_group]:

                self.parents[b_group] = a_group

            elif self.ranks[b_group] > self.ranks[a]:

                self.parents[a_group] = b_group

            else:

                self.parents[b_group] = a_group

                self.ranks[a_group] += 1

            self.groups -= 1

    def is_connected(self, a, b):

        return self.find(a) == self.find(b)

def main():

    for unionfind_class in [UnionFind1]:

        uf = unionfind_class(10)

        print(uf.groups)

        uf.union(0, 1)
        uf.union(2, 3)
        uf.union(4, 5)

        print(uf.is_connected(0, 3))

        uf.union(1, 2)

        print(uf.is_connected(0, 3))

        print(uf.is_connected(0, 2))

        print(uf.is_connected(2, 5))

        print(uf.groups)

        print("------")

if __name__ == '__main__':
    main()
